### PR TITLE
Added argument to get_distance_matrix to change the 'max_nbytes' parameter of joblib.Parallel

### LINKED ISSUE #2060: Parallelisation failed- get_distance_matrix()
**Expected behavior**
Parallel computation (with 2 or more cores) of a rmsd matrix with `MDAnalysis.analysis.encore.confdistmatrix.get_distance_matrix()`.

**Actual behavior**
Running the following command with my PSF and DCD, I get this error.
```python
import MDAnalysis as mda
import MDAnalysis.analysis.encore as encore

u1 = mda.Universe('my.psf', 'my.dcd', in_memory=True, in_memory_step=10)

rmsdMatrix = encore.confdistmatrix.get_distance_matrix(
        u1, selection='name CA', superimpose=True, n_jobs=-1, weights='mass', metadata=True, verbose=True)
```

Output:
```python
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
---------------------------------------------------------------------------
_RemoteTraceback                          Traceback (most recent call last)
_RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/jarvis/anaconda2/envs/py3/lib/python3.6/site-packages/joblib/externals/loky/process_executor.py", line 425, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
  File "/home/jarvis/anaconda2/envs/py3/lib/python3.6/site-packages/joblib/_parallel_backends.py", line 561, in __call__
    return self.func(*args, **kwargs)
  File "/home/jarvis/anaconda2/envs/py3/lib/python3.6/site-packages/joblib/parallel.py", line 261, in __call__
    for func, args, kwargs in self.items]
  File "/home/jarvis/anaconda2/envs/py3/lib/python3.6/site-packages/joblib/parallel.py", line 261, in <listcomp>
    for func, args, kwargs in self.items]
  File "/home/jarvis/anaconda2/envs/py3/lib/python3.6/site-packages/MDAnalysis/analysis/encore/confdistmatrix.py", line 244, in set_rmsd_matrix_elements
    coords[j].shape[0], weights, sumweights)
ValueError: assignment destination is read-only
"""

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
<ipython-input-31-28b74f5d43a7> in <module>()
      1 rmsdMatrix = encore.confdistmatrix.get_distance_matrix(
----> 2         u1, selection='name CA', superimpose=True, n_jobs=-1, weights='mass', metadata=True, verbose=True)

~/anaconda2/envs/py3/lib/python3.6/site-packages/MDAnalysis/analysis/encore/confdistmatrix.py in get_distance_matrix(ensemble, selection, load_matrix, save_matrix, superimpose, superimposition_subset, weights, n_jobs, verbose, *conf_dist_args, **conf_dist_kwargs)
    361                                                         weights=weights,
    362                                                         n_jobs=n_jobs,
--> 363                                                         verbose=verbose)
    364 
    365         logging.info("    Done!")

~/anaconda2/envs/py3/lib/python3.6/site-packages/MDAnalysis/analysis/encore/confdistmatrix.py in conformational_distance_matrix(ensemble, conf_dist_function, selection, superimposition_selection, n_jobs, pairwise_align, weights, metadata, verbose)
    175         weights,
    176         fitting_coordinates,
--> 177         subset_weights) for element in indices)
    178 
    179 

~/anaconda2/envs/py3/lib/python3.6/site-packages/joblib/parallel.py in __call__(self, iterable)
    992 
    993             with self._backend.retrieval_context():
--> 994                 self.retrieve()
    995             # Make sure that we get a last message telling us we are done
    996             elapsed_time = time.time() - self._start_time

~/anaconda2/envs/py3/lib/python3.6/site-packages/joblib/parallel.py in retrieve(self)
    895             try:
    896                 if getattr(self._backend, 'supports_timeout', False):
--> 897                     self._output.extend(job.get(timeout=self.timeout))
    898                 else:
    899                     self._output.extend(job.get())

~/anaconda2/envs/py3/lib/python3.6/site-packages/joblib/_parallel_backends.py in wrap_future_result(future, timeout)
    513         AsyncResults.get from multiprocessing."""
    514         try:
--> 515             return future.result(timeout=timeout)
    516         except LokyTimeoutError:
    517             raise TimeoutError()

~/anaconda2/envs/py3/lib/python3.6/site-packages/joblib/externals/loky/_base.py in result(self, timeout)
    429                 raise CancelledError()
    430             elif self._state == FINISHED:
--> 431                 return self.__get_result()
    432             else:
    433                 raise TimeoutError()

~/anaconda2/envs/py3/lib/python3.6/site-packages/joblib/externals/loky/_base.py in __get_result(self)
    380     def __get_result(self):
    381         if self._exception:
--> 382             raise self._exception
    383         else:
    384             return self._result

ValueError: assignment destination is read-only
```
**Code to reproduce the behavior**

I cannot reproduce the same error if I use the test file or a different number of cores. And nothing changed in my enviroment between the two computation.

I also checked my data, but they are correctly read by VMD and does not appear to be corrupted.
Using only one core, everything works fine.

```python
import MDAnalysis as mda
import MDAnalysis.analysis.encore as encore
from MDAnalysisTests.datafiles import PSF, DCD

u = mda.Universe(PSF, DCD, dt=1, in_memory=True, in_memory_step=2 )

rmsdMatrix = encore.confdistmatrix.get_distance_matrix(
u, selection='name CA', superimpose=True, n_jobs=2, weights='mass', metadata=True, verbose=True)
```
Output:
```python
[Parallel(n_jobs=2)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=2)]: Done 1225 out of 1225 | elapsed:    1.1s finished
```

**Currently version of MDAnalysis**

- Which version are you using? '0.18.0'
- Which version of Python (`python -V`)? Python 3.6.5 | packaged by conda-forge | (default, Apr  6 2018, 13:39:56) on jupyter-notebook
- Which operating system? ubuntu 16.04

### PR DESCRIPTION
**Pull request by mjtadema**
By default this is set to '1M', but raises a 'destination is read-only' exception in some cases. 
While trying to debug this issue i came across [this](https://github.com/scikit-learn/scikit-learn/issues/5956) issue in scikit-learn which also uses joblib. 
They suggest to increase the 'max_nbytes' parameter of joblist.Parallel as needed, which allows for larger temporary files. I'm not sure why, but this fixes the problem for me.
I thought it'd be nice to add this as an argument to get_distance_matrix so max_nbytes can be increased if necessary.

Fixes #2060 

Changes made in this Pull Request:
 - Added argument to get_distance_matrix to allow changing the 'max_nbytes' parameter of joblib.Parallel

PR Checklist
------------
 - [x] Tests?
 - [ ] Docs?
 - [x] CHANGELOG updated?
 - [X] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/analysis/encore/confdistmatrix.py [hunk 1]
```diff
@@ -56,7 +56,7 @@ class to compute an RMSD matrix in such a way is also available.
 
 def conformational_distance_matrix(ensemble,
                                    conf_dist_function, selection="",
-                                   superimposition_selection="", n_jobs=1, pairwise_align=True, weights='mass',
+                                   superimposition_selection="", n_jobs=1, max_nbytes='1M', pairwise_align=True, weights='mass',
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/encore/confdistmatrix.py]: document the parameter and include a link to https://joblib.readthedocs.io/en/latest/generated/joblib.Parallel.html

File: package/MDAnalysis/analysis/encore/confdistmatrix.py [hunk 2]
```diff
@@ -256,6 +256,7 @@ def get_distance_matrix(ensemble,
                         superimposition_subset="name CA",
                         weights='mass',
                         n_jobs=1,
+                        max_nbytes='1M',
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/encore/confdistmatrix.py]: document (as above)

File: package/MDAnalysis/analysis/encore/confdistmatrix.py (hunk 1 above)
Comment (mjtadema) [reply package/MDAnalysis/analysis/encore/confdistmatrix.py]: Will this generate the correct documentation?

**Comment by mtiberti** [comment]
Hi, thanks a lot for working on this. I had tried a similar solution in the issue by setting `max_nbytes=None`, but I was getting really strange numbers, I think because the matrix wasn't properly shared among the workers by joblib. I seem to have a similar behaviour with your fork when setting `n_job`, can you double check?

**Comment by mjtadema** [comment]
You are right, after some testing with a smaller dataset that would finish within reasonable time, it does produce nonsense.
Do you know whether this matrix sharing is also broken when using the default of 1M on a smaller dataset that won't trigger the error?
In any case I do have some good news. While trying to compute this matrix i had some horrible bottleneck somewhere, which led me to experiment with the different parallelization backends. 
I ended up trying the 'threading' backend, and it works great for me, produces the actual results and is much faster. The threading backend doesn't need to write these tempfiles to share memory so that is not an issue either. And the global interpreter lock is not an issue because the heavy operations are in numpy and they are not blocked.
Maybe a better fix for this particular problem would be to add a `prefer="threads"` to the call to joblist.Parallel. This would then still allow the user to change any parameters via the parallel_backend context manager, which seems to be their preferred way of doing things.
Unless you see a problem with this?

**Comment by mtiberti** [comment]
Thanks - that would be a great solution! I'm wondering is whether the backend release the global interpreter lock (i.e. does it run effectively on multiple cores)? I can't find anywhere that says it does in the generic case. But if it is faster anyway (I'm wondering if joblib's overhead is worth it at this point) we can certainly leave it in as an option.

Alternatively, I just found the `require='sharedmem'` option of `joblib.Parallel` that seems to do the trick. I ran a quick test with `n_jobs=4` and `max_nbytes=None` and I get the same matrix. I'm not sure it's something recent, I'd never seen it before in my previous searches

**Comment by orbeckst** [comment]
If you decide to default to a new option in joblib then increase the required version of joblib in `setup.py` and in the `.travis.yml` file.

**Comment by orbeckst** [comment]
@mtiberti could you please review this PR? Thanks!

**Comment by orbeckst** [comment]
@mjtadema could you please address @mtiberti 's comments? It would be great to get this fixed!

**Comment by mjtadema** [comment]
> @mjtadema could you please address @mtiberti 's comments? It would be great to get this fixed!

I've made the changes.
In my limited testing it seems to work as intended and produce the correct results.
Can you test it with a larger dataset @mtiberti ?

**Comment by mtiberti** [comment]
Thanks @mjtadema ! I've a couple more tests with beefier datasets and few different number of cores and I agree it works as expected. I have added the required minimum version of `joblib` in `.travis.yml`(hopefully correctly!)

**Comment by orbeckst** [comment]
@mtiberti you should be able to merge this one yourself – please do it! (If you want someone else to give it a last look over then ping the other core devs but otherwise: you have the power :-)).

@mjtadema and @mtiberti thank you for working on the PR and seeing it through to the end. Sorry I didn't see sooner that it was just waiting for a merge.

**Comment by orbeckst** [comment]
@mjtadema I sent you an invite to become a member of the MDAnalysis contributors.

**Comment by mtiberti** [comment]
@orbeckst thanks - sorry for the merge-shyness :) I'll do it then, since tests pass it should be alright

**Comment by mjtadema** [comment]
> @mtiberti you should be able to merge this one yourself – please do it! (If you want someone else to give it a last look over then ping the other core devs but otherwise: you have the power :-)).
> 
> @mjtadema and @mtiberti thank you for working on the PR and seeing it through to the end. Sorry I didn't see sooner that it was just waiting for a merge.

No worries, happy to contribute!

**Pull request merged**